exchanges/kucoin/kucoin_websocket.go

```diff
@@ -0,0 +1,1870 @@
+		marketOrderbookLevel2Channels, marketOrderbookLevel2to5Channel,
+		marketOrderbokLevel2To50Channel, marketCandlesChannel,
+		marketMatchChannel, indexPriceIndicatorChannel,
+		markPriceIndicatorChannel, marginFundingbookChangeChannel,
+		privateSpotTradeOrders, accountBalanceChannel,
+		marginPositionChannel, marginLoanChannel,
+		spotMarketAdvancedChannel:
+		return asset.Spot, nil
+	default:
+		return asset.Empty, errors.New("channel not supported")
+	}
+}
+
+// GenerateDefaultSubscriptions Adds default subscriptions to websocket.
+func (ku *Kucoin) GenerateDefaultSubscriptions() ([]stream.ChannelSubscription, error) {
+	defaultSubscriptionsChoice.Do(func() {
+		if ku.CurrencyPairs.IsAssetEnabled(asset.Spot) == nil || ku.CurrencyPairs.IsAssetEnabled(asset.Margin) == nil {
+			defaultSubscriptionChannels = append(defaultSubscriptionChannels,
+				marketTickerChannel,
+				marginFundingbookChangeChannel,
+				marketCandlesChannel,
+				marketOrderbokLevel2To50Channel)
+		}
+		if ku.CurrencyPairs.IsAssetEnabled(asset.Margin) == nil {
+			defaultSubscriptionChannels = append(defaultSubscriptionChannels,
+				marginFundingbookChangeChannel)
+		}
+		if ku.CurrencyPairs.IsAssetEnabled(asset.Futures) == nil {
+			defaultSubscriptionChannels = append(defaultSubscriptionChannels,
+				futuresTickerV2Channel,
+				futuresOrderbookLevel2Depth50Channel)
+		}
+	})
+	channels := defaultSubscriptionChannels
+	subscriptions := []stream.ChannelSubscription{}
+	if ku.Websocket.CanUseAuthenticatedEndpoints() {
+		if ku.CurrencyPairs.IsAssetEnabled(asset.Spot) == nil {
+			channels = append(channels,
+				accountBalanceChannel,
+			)
+		}
+		if ku.CurrencyPairs.IsAssetEnabled(asset.Margin) == nil {
+			channels = append(channels,
+				marginPositionChannel,
+				marginLoanChannel,
+			)
+		}
+		if ku.CurrencyPairs.IsAssetEnabled(asset.Futures) == nil {
+			channels = append(channels,
+				// futures authenticated channels
+				futuresTradeOrdersBySymbolChannel,
+				futuresTradeOrderChannel,
+				futuresStopOrdersLifecycleEventChannel,
+				futuresAccountBalanceEventChannel)
+		}
+	}
+	var err error
+	var spotPairs currency.Pairs
+	if ku.CurrencyPairs.IsAssetEnabled(asset.Spot) == nil {
+		spotPairs, err = ku.GetEnabledPairs(asset.Spot)
+		if err != nil {
+			return nil, err
+		}
+	}
+	var marginPairs currency.Pairs
+	if ku.CurrencyPairs.IsAssetEnabled(asset.Margin) == nil {
+		marginPairs, err = ku.GetEnabledPairs(asset.Margin)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// jointSpotMarginPairs used to hold spot and margin currency pairs to be used to subscribe orderbook, tickers, and candlestick channels.
+	var jointSpotMarginPairs currency.Pairs
+	format, err := ku.GetPairFormat(asset.Spot, true)
+	if err == nil {
+		var spotMarginPairsDiff currency.PairDifference
+		if spotMarginPairsDiff, err = spotPairs.FindDifferences(marginPairs, format); err == nil {
+			jointSpotMarginPairs = append(jointSpotMarginPairs, spotMarginPairsDiff.New...)
+		}
+	}
```

Comment:
This does not work. Having the same currency in both spot and margin results in neither pair being subscribed and instead an empty request being sent:
`Kucoin websocket connection: message received: {"id":"143907219","type":"error","code":400,"data":"topic /market/ticker: is invalid"}`
 The reliance on using err == nil {} and not capturing errors is also of concern, but not the main thing I'm focussing on just yet.

A subscription map so that only one currency pair may be subscribed to + your recent work on ensuring subscription data populates to the appropriate asset/currency pair may be of help